File: csharp/creational/rare_singleton.cs
namespace Gof.ModernCSharp.Creational;

// Singleton ensures repeated access returns the same shared instance.
// Modern C# note: prefer DI lifetimes such as AddSingleton over static global
// access. Direct Singleton classes hide dependencies and complicate tests.
public static class RareSingleton
{
    sealed class Settings
    {
        public static Settings Instance { get; } = new();
        public string Theme { get; set; } = "light";
        private Settings() { }
    }

    public static string Run()
    {
        Settings.Instance.Theme = "dark";
        return Settings.Instance.Theme;
    }
}

File: py/creational/rare_singleton.py
# Singleton ensures repeated construction returns the same shared instance.
# Modern Python note: prefer module-level objects or explicit dependency injection.
# Direct Singleton classes hide dependencies and can make tests harder to isolate.
class Settings:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.theme = "light"
        return cls._instance


if __name__ == "__main__":
    first = Settings()
    second = Settings()
    second.theme = "dark"
    print(first is second)
    print(first.theme)
